### PR TITLE
Fix profilers plot not being aligned to context manager enter time

### LINKED ISSUE #9478: Diagnostics graphs are not time aligned
**What happened**:

While working with some dask-based code I used the `Profiler`, `ResourceProfiler`, and `CacheProfiler` to debug which tasks were not as dask-friendly as I'd like. The code being run takes a non-negligible amount of time to generate the dask array. This means the resource profiler is collecting information for a while before the other profiles have seen their first task being computed/cached. This results in the resource (CPU/memory) plot not being aligned with the task-based Profiler and CacheProfiler plots. The latter plots are plotted on the left even though they didn't start at that point in time.

If you run the below code and open the two HTML files it creates (you might need to zoom out on the plots), you'll see that the "nostartertask" plot shows the dask tasks executing at time 0 even though those tasks weren't executed until 5 seconds later. The "startertask" plots show that if the "starter task" is used as a "reference" for time 0 then things line up a little better.

### No Starter Task

image

### With Starter Task

image

**What you expected to happen**:

When visualizing Profilers/diagnostic tools together, they should be aligned to the same time range.

**Minimal Complete Verifiable Example**:

```python
import time

import dask.array as da
from dask.diagnostics import ResourceProfiler, Profiler, CacheProfiler, visualize


if __name__ == "__main__":
    with Profiler() as prof, ResourceProfiler() as rprof, CacheProfiler() as cprof:
        # da.random.random((100, 100), chunks=10).compute()
        time.sleep(5)
        da.random.random((100, 100), chunks=10).compute()
    filename = '/tmp/debug_diagnostics_nostartertask.html'
    visualize([rprof, prof, cprof], filename=filename, show=False)
    print(f"file://{filename}")

    with Profiler() as prof, ResourceProfiler() as rprof, CacheProfiler() as cprof:
        da.random.random((100, 100), chunks=10).compute()
        time.sleep(5)
        da.random.random((100, 100), chunks=10).compute()
    filename = '/tmp/debug_diagnostics_startertask.html'
    visualize([rprof, prof, cprof], filename=filename, show=False)
    print(f"file://{filename}")


```

**Anything else we need to know?**:

**Environment**:

- Dask version: 2022.4.0
- Python version: 3.9
- Operating System: PopOS (Ubuntu)
- Install method (conda, pip, source): conda-forge

Edit: Just tested with 2022.9.0 and same results.

### PR DESCRIPTION
**Pull request by djhoese**
As mentioned in #9478, the Profiler and CacheProfiler plots X (time) range is set based on the first and last element in their results. Since these elements are based on when dask tasks are computed you can run into alignment issues when combining plots with the ResourceProfiler which starts collecting data when its context manager is entered.

This PR fixes this by adding a `start_time` and `end_time` property to all of the profiles that are set in `__enter__` and `__exit__`. They are then used for plotting which makes sure the plots are generally aligned with in a couple milliseconds depending on how fast all the context manager's `__enter__` methods are called and that they include all the data for the entire time spent inside the context.

Some questions, todos, and other context:

1. In general my opinion is that if you don't want something to show up in the plots/profilers then it shouldn't happen inside the `with` block. Without this PR I was having the opposite issue where stuff was happening in the context that wasn't being shown in the default x-range of the plots (ex. resource profiler going on longer as something is being saved to disk).
2. I made the start_time and end_time public (no `_`) because I don't really like writing tests that check private properties and I thought they could be useful to users.
3. This PR assumes the plotting functions used by each profiler are not used outside of the `visualize` function and that backwards compatibility isn't an issue.
4. ~I'll see if I can add some checks to the plot tests too.~ The plots are all relative time on their X axis so without diving into private properties on the bokeh objects I'm not sure how to check that the plot looks as expected.

Here is what the plots look like for the "no starter task" case from the issue with the default view/zoom with this PR:

image

Note the extra space on the right of the plots. I think this is due to the time it takes to kill the ResourceProfiler's subprocess. If the `end_time` was set before the process was terminated this would probably get rid of this. I'm not sure if it matters in a real world case where these graphs will be larger than 5 seconds in a lot of cases.

- [x] Closes #9478
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by djhoese** [comment]
Is there any way I can get someone to look at this? It includes relatively simple changes, but is a pretty big quality of life fix for me.

File: dask/diagnostics/tests/test_profiler.py [hunk 1]
```diff
@@ -270,6 +277,8 @@ def test_resource_profiler_plot():
     rprof.clear()
     for results in [[], [(1.0, 0, 0)]]:
         rprof.results = results
+        rprof.start_time = 0.0
+        rprof.end_time = 1.0
```
Comment (jrbourbeau) [inline dask/diagnostics/tests/test_profiler.py:281]: Why is this needed?

**Review by jrbourbeau — COMMENTED** [review]
Thanks for the nudge @djhoese, apologies for the delayed response. The changes here generally seem good to me. Just one minor question

File: dask/diagnostics/tests/test_profiler.py (hunk 1 above)
Comment (djhoese) [reply dask/diagnostics/tests/test_profiler.py:281]: Forgive me if this is incorrect, it's been a while since I looked at this. But from what I see in this PR, the `rprof.clear()` call before this `for` loop sets `start_time` and `end_time` to `None` (which matches what was in the clear method before this PR). Now that `start_time` and `end_time` are actually used for something in the visualization they can't be `None` because of the arithmetic operations being done with them.

...yeah, this line (218 of profile.py):

https://github.com/dask/dask/pull/9739/files#diff-d1ec632f95ddf2554df9d54f70b8980b9f24699be167aa97c8b724c18a7da88dR218

Previously only `results` were sent to the visualize function for the ResourceProfiler.

Comment (quasiben) [reply dask/diagnostics/tests/test_profiler.py:281]: end_time here is arbitrary picked just to visualize the for the test, correct ?  It could have also been a call to the current time?

**Comment by quasiben** [comment]
I merged in main for another round of testing.  Once tests pass I think we should be good to merge

File: dask/diagnostics/tests/test_profiler.py (hunk 1 above)
Comment (djhoese) [reply dask/diagnostics/tests/test_profiler.py:281]: Yes. I think that's right.

Comment (djhoese) [reply dask/diagnostics/tests/test_profiler.py:281]: Looking it over more, this start and end setting could be seen as the initialization of fake results/usage of the profiler. I suppose this could be made more clear with a fixture or a helper context manager.

**Comment by quasiben** [comment]
Thanks @djhoese .  Merging in
